src/extractor/guidance/turn_handler.cpp

```diff
@@ -470,125 +527,133 @@ Intersection TurnHandler::assignRightTurns(const EdgeID via_edge,
+            const int right_index = intersection.rend() - right - 1;
+            const auto forward_right = intersection.begin() + right_index;
+            // find the leftmost road that might be part of a fork
+            const auto left = findOutermostForkCandidate(straightmost, intersection.end());
+
+            // if the leftmost and rightmost roads with the conditions above are the same
+            // or if there are more than three fork candidates
+            // they cannot be fork candidates
+            if (forward_right < left && left - forward_right + 1 <= 3)
+            {
+                return Fork(forward_right, left);
+            }
         }
     }
-    if (best_deviation <= NARROW_TURN_ANGLE)
-    {
-        std::size_t left = best, right = best;
-        while (
-            left + 1 < intersection.size() &&
-            (angularDeviation(intersection[left + 1].angle, STRAIGHT_ANGLE) <= NARROW_TURN_ANGLE ||
-             (angularDeviation(intersection[left].angle, intersection[left + 1].angle) <=
-                  NARROW_TURN_ANGLE &&
-              angularDeviation(intersection[left].angle, STRAIGHT_ANGLE) <= GROUP_ANGLE)))
-            ++left;
-        while (
-            right > 1 &&
-            (angularDeviation(intersection[right - 1].angle, STRAIGHT_ANGLE) <= NARROW_TURN_ANGLE ||
-             (angularDeviation(intersection[right].angle, intersection[right - 1].angle) <
-                  NARROW_TURN_ANGLE &&
-              angularDeviation(intersection[right - 1].angle, STRAIGHT_ANGLE) <= GROUP_ANGLE)))
-            --right;
-
-        if (left == right)
-            return std::make_pair(std::size_t{0}, std::size_t{0});
-
-        const bool valid_indices = 0 < right && right < left;
+    return boost::none;
+}
+
+// check if the fork candidates (all roads between left and right) and the
+// incoming edge are compatible by class
+bool TurnHandler::isCompatibleByRoadClass(const Intersection &intersection, const Fork fork) const
+{
+    const auto via_class = node_based_graph.GetEdgeData(intersection[0].eid).road_classification;
+
+    // if any of the considered roads is a link road, it cannot be a fork
+    // except if rightmost fork candidate is also a link road
+    const auto is_right_link_class =
+        node_based_graph.GetEdgeData((*fork.right).eid).road_classification.IsLinkClass();
+    if (!std::all_of(fork.right + 1, fork.left + 1, [&](ConnectedRoad &road) {
+            return is_right_link_class ==
+                   node_based_graph.GetEdgeData(road.eid).road_classification.IsLinkClass();
+        }))
+    {
+        return false;
+    }
+
+    return std::all_of(fork.right, fork.left + 1, [&](ConnectedRoad &base) {
+        const auto base_class = node_based_graph.GetEdgeData(base.eid).road_classification;
+        // check that there is no turn obvious == check that all turns are non-onvious
+        return std::all_of(fork.right, fork.left + 1, [&](ConnectedRoad &compare) {
```

Comment:
Does the second loop have to exclude the road from the first road? Otherwise you're doing a self-comparison?